File: app/src/main/java/com/example/wagtailjournal/Musuko.java
package com.example.wagtailjournal;

import java.io.File;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;

public class Musuko {
    Date timestamp;

    public File getFile(File dir) {
        if (timestamp == null)
            updateTimestampByFile();
        String filename = getDateFormat().format(timestamp);
        return new File(dir, filename);
    }

    private SimpleDateFormat getDateFormat() {
        return new SimpleDateFormat("yyyyMMddHHmmss'.txt'");
    }

    public void updateTimestampByFile() {
        timestamp = new Date();
    }

    public void updateTimestampByFile(File file) throws ParseException {
        if (file == null) {
            updateTimestampByFile();
            return;
        }
        String filename = file.getName();
        timestamp = parseDate(filename);
    }

    private Date parseDate(String date) throws ParseException {
        try {
            return getDateFormat().parse(date);
        } catch (ParseException e) {
            return new SimpleDateFormat("yyyy-MM-dd'.txt'").parse(date);
        }
    }
}

File: app/src/main/java/com/github/tkojitu/wagtailjournal/Musuko.kt
package com.github.tkojitu.wagtailjournal

import java.io.File
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Date

class Musuko {
    var timestamp: Date? = null

    fun getFile(dir: File?): File {
        if (timestamp == null) updateTimestampByFile()
        val filename = dateFormat.format(timestamp)
        return File(dir, filename)
    }

    private val dateFormat: SimpleDateFormat
        get() = SimpleDateFormat("yyyyMMddHHmmss'.txt'")

    fun updateTimestampByFile() {
        timestamp = Date()
    }

    @Throws(ParseException::class)
    fun updateTimestampByFile(file: File?) {
        if (file == null) {
            updateTimestampByFile()
            return
        }
        val filename = file.name
        timestamp = parseDate(filename)
    }

    @Throws(ParseException::class)
    private fun parseDate(date: String): Date {
        return try {
            dateFormat.parse(date)
        } catch (e: ParseException) {
            SimpleDateFormat("yyyy-MM-dd'.txt'").parse(date)
        }
    }
}
